Moment of Truth: this entire school year's worth of changes running on *every* mcu.

ds196 opened the pull request:
This home-grown PlatformIO library allows *every single one of our 8 microcontrollers on Clucky to function*. Dis shit better be **spotless**. Critique, comment, question, and nit-pick to your heart's desire--show me your worst. I want **everyone to sign off** on the changes.

### A brief summary of changes made in this pull request:

 - Added pins and constants for six new PCBs
 - Completely re-wrote AstraREVCAN
 - Integrated AstraREVCAN into AstraMotors
 - Added AstraVicCAN.h, supporting ***all microcontroller communication*** on Clucky :fearful:
 - Added a template main.cpp
 - Revamped documentation

Charles-Marmann committed:
Update CORE.h

Charles-Marmann committed:
Start creating the cpp and h file

Charles-Marmann committed:
Update AstraCAN2.cpp

Charles-Marmann committed:
Update all old functions

Charles-Marmann committed:
Define pins in core.h

Charles-Marmann committed:
Update AstraMotors.cpp

Charles-Marmann committed:
Update AstraMotors.h

Charles-Marmann committed:
a

Charles-Marmann committed:
Update AstraCAN2.h

Charles-Marmann committed:
a

Charles-Marmann committed:
Update AstraCAN2.h

These libraries were already included in the old CAN library.

Charles-Marmann committed:
Comment out the old CAN

I don't want to get rid of these files, but they are causing errors, if I remember I will take care of them.

Charles-Marmann committed:
Update AstraCAN2.h

Charles-Marmann committed:
HOLLY SHIT IM DUMB AS A MOTHER FUCKER

Charles-Marmann committed:
Update AstraCAN2.cpp

Charles-Marmann committed:
Update AstraCAN2.cpp

Charles-Marmann committed:
Forgot to change this

Charles-Marmann committed:
Add a second set of pin constants for the motor MC

ds196 committed:
Move AstraCAN2 functionality to AstraCAN

While before, AstraCAN was for Teensy and AstraCAN2 was for ESP32, now AstraCAN  will automatically adjust to either ESP32 or Teensy, using the correct library while keeping its interface the same with main.cpp.

ds196 committed:
Add canObject (Can0) to AstraMotors class

ds196 committed:
Add setBrake() method to AstraMotors

Abstracts away setParameter, should be less confusing and easier to use.

ds196 committed:
Format AstraMotors.cpp

ds196 committed:
Add sendDuty() method to AstraMotors

ds196 committed:
Add note that sendDuty(float) bypasses acceleration

Charles-Marmann committed:
Update CORE.h

ds196 committed:
Change AstraMotors.canObject to a pointer

Avoids possible issues with canObject simply being a copy of Can0, now each AstraMotors object holds a pointer to Can0 and uses that instead.

ds196 committed:
Flip CTX and RTX pins CORE.h

ds196 committed:
Change AstraCAN2 deprecation directive to warning

ds196 committed:
Move testbed pins to TESTBED.h from CORE.h

Testbed projects now need to use "project/TESTBED.h" instead of "project/CORE.h". There may be a platform.io macro involved with this in the future.

ds196 committed:
Remove AstraCAN2

ds196 committed:
Add AstraMotors::identify() to run IdentifyDevice()

ds196 committed:
Start adding comments and reorganize AstraMotors.cpp

ds196 committed:
Refactor: change AstraFCAN to AstraCAN

ds196 committed:
Add AstraMotors.accelerate() to run UpdateForAcceleration and sendDuty in one function

Charles-Marmann committed:
Update AstraCAN.cpp

Charles-Marmann committed:
Update AstraCAN.cpp

Charles-Marmann committed:
Update AstraCAN.cpp

Charles-Marmann committed:
have sendDutyCycle return the pointer to msg.data so that we can print it in main (Insanity)

Charles-Marmann committed:
IT NEVER ENDS

Trying to use Serial.print in this, I don't think it will work, but there is only one way to find out.

Charles-Marmann committed:
IT DOES WORK!!!

Adding more Serial.print() to check what the hell is going to with this bullshit.

Charles-Marmann committed:
For some reason it doesn't like this

Charles-Marmann committed:
Getting warmer

Charles-Marmann committed:
EDGING!

Charles-Marmann committed:
Update AstraMotors.cpp

Charles-Marmann committed:
I think I found the issue!

Charles-Marmann committed:
Winning!

Charles-Marmann committed:
Winning Real

Charles-Marmann committed:
Update AstraMotors.cpp

Charles-Marmann committed:
Update AstraMotors.cpp

Charles-Marmann committed:
Remove print statements + other

Revert some changes that were made when I was trying to find out why the floats were getting rounded.

ds196 committed:
Fix map() in AstraMotors.cpp

Used long, so decimal places were dropped. Now uses double.

ds196 committed:
Fix Template.cpp to be microcontroller-agnostic

ds196 committed:
Add PIN_EF_MOTOR to digit for end effector motor PWM

ds196 committed:
Initial commit for AstraNP.h

Handles control of Feather NeoPixels for showing Feather status

ds196 committed:
Add AstraREVCAN

ds196 committed:
Fix up AstraREVCAN

Implement heartbeat and identify, fix (hopefully) setParameter, change names of all the CAN functions, and implement all behavior of AstraCAN to replace it

ds196 committed:
Disable AstraCAN

Now requires OLD_ASTRACAN_ENABLE macro for usage

ds196 committed:
Move AstraMotors to new AstraREVCAN

ds196 committed:
Uncomment sendHeartbeat() for Teensy in AstraREVCAN

ds196 committed:
Fix parameter type for kIdleMode

ds196 committed:
Add CAN_sendSpeed for motor speed control

ds196 committed:
Add motorStatus1 for motor temperature and current reading

ds196 committed:
Prepare project headers for future ESP32 boards

ds196 committed:
Add motor IDs to TESTBED.h

ds196 committed:
Move AstraMisc.cpp into AstraMisc.h

Not necessary at all to keep these seperate, all function code can go in AstraMisc.h

ds196 committed:
Deprecate parseInput() with `const char delm`, use version without

This function should use the macro also defined in AstraMisc

ds196 committed:
Add checkArgs() to check for correct number of arguments in command

ds196 committed:
Fix setParameter() so brake mode works

ds196 committed:
fix: CAN_sendHeartbeat uses different format then other commands

ds196 committed:
refactor AstraMotors

rename set to target in setMotorSpeed and setDutyCycle, refactored UpdateForAcceleration, and reorganized class variables

ds196 committed:
feat: add debug log for CAN_sendPacket()

ds196 committed:
Add CAN_enumerate to ping motor controllers

ds196 committed:
refactor: reorganize AstraREVCAN

* ESP32 and Teensy now use all the same functions except for interfacing with their CAN library
* Backwards compatible functions are available for both ESP32 and Teensy, and they use the new functions
* Should be easier to read

ds196 committed:
refactor: add note for CAN_sendPacket

ds196 committed:
docs: reorganize main README and add Overview section

ds196 committed:
docs: add items to Overview section

I already added these?!?!??!?

ds196 committed:
docs: start readme for project headers

ds196 committed:
docs: finishing explaining the submodules

ds196 committed:
docs: add TOC and responsible people

ds196 committed:
feat: add function to parse status1 msgs

ds196 committed:
feat: add preliminary motor velocity parsing

ds196 committed:
docs: added "updating your libraries" section

ds196 committed:
style: change functions back to inline and move to AstraMotors.h

ds196 committed:
feat: add func to parse status2 msg

ds196 committed:
style: revert to old CAN functions in AstraMotors for easier debugging

ds196 committed:
fix: pain

ds196 committed:
refactor: move status parsing code from AstraREVCAN to AstraMotors

ds196 committed:
style: add more default args to AstraMotors constructor

ds196 committed:
feat: add gearBox to AstraMotors

ds196 committed:
feat: add function to turn motor by degrees

ds196 committed:
feat: detect overshoot in rotating to pos

ds196 committed:
test: add debug messages and tune turn speed

ds196 committed:
feat: add WHEEL_CIRCUMFERENCE to standardize between core and testbed

ds196 committed:
style!: change stopTurn() to stop() in AstraMotors

ds196 committed:
feat: add AstraMotors.direction() to check direction of current rotation

Returns 1 for positive rotation, -1 for negative rotation, and 0 if not spinning. Accounts for speed/dutycycle control mode.

ds196 committed:
feat: integrate direction() into position check to account for control mode

ds196 committed:
feat: account for control mode in AstraMotors.turnByDeg()

ds196 committed:
style: remove compiler warning in AstraMotors.direction()

ds196 committed:
feat: add getParameter() and ctrlType

ds196 committed:
style!: use enum class for controlmode

ds196 committed:
feat: add AstraMotors.isRotToPos()

ds196 committed:
feat!: add CAN send smart velocity and position, rename sendSpeed to sendVelocity

ds196 committed:
feat: setParameter() adapts to param type

ds196 committed:
feat: add printREVFrame() for debugging

ds196 committed:
style!: fix parameter status enum name

ds196 committed:
feat: add AstraMotors::sendSpeed() to make speed control mode work

ds196 committed:
feat: add status0

ds196 committed:
feat: add speed control mode condition to UpdateForAcceleration()

ds196 committed:
style: add standard code layout to Template

ds196 committed:
feat: add getGearBox() to AstraMotors

ds196 committed:
style: make column limit more lenient in .clang-format

ds196 committed:
feat: add standard sections for serial commands to Template.cpp

ds196 committed:
feat: add Timer struct

ds196 committed:
feat: add hwPinToGPIO() for pico

ds196 committed:
feat: add stopwatch

ds196 committed:
remove AstraArm (moved to arm embedded pio project)

ds196 committed:
feat: add changeNeopixel()

ds196 committed:
feat: add pins to CORE.h

ds196 committed:
fix: AstraNP now correctly sets NeoPixel

ds196 committed:
refactor: move generic CAN code to AstraCAN from AstraREVCAN

ds196 committed:
feat: add printCANframe()

ds196 committed:
feat: introduce AstraVicCAN.h for inter-mcu comms

ds196 committed:
style: disallow one-line enum declarations

ds196 committed:
feat: layout handling built-in CAN commands

ds196 committed:
feat: add pins for digit

ds196 committed:
refactor: move map() from AstraMotors to AstraMisc

ds196 committed:
style: add MCU section comments to project headers

ds196 committed:
fix: now compiles
Wasn't done schizo coding AstraVicCAN

lover304 committed:
Added FAERIE pins

lover304 committed:
Merge branch 'esp32-dev' of https://github.com/SHC-ASTRA/rover-Embedded-Lib into esp32-dev

ds196 committed:
fix: now compiles without AstraREVCAN

Not tested tho

ds196 committed:
feat: reorganize AstraVicCAN, add VicCanController

ds196 committed:
feat: add AstraVicCanController::relayFromSerial()

ds196 committed:
feat: add VicCan respond() with single integer

ds196 committed:
feat: add command to parse vic CAN data

ds196 committed:
feat: parseData() infers type, added rest of datatypes

ds196 committed:
feat: add repond() for all data types

ds196 committed:
refactor: move viccan data encoding to encodeData() from respond()

Will be used by explicit send function aswell as respond()

ds196 committed:
style: move readyTxFrame() to make it easier to find

ds196 committed:
feat: add send() for vic can one-way telemetry

ds196 committed:
feat: add viccan commands for sensor data

ds196 committed:
fix: added "MCU_" to mcu id's for vic can

ds196 committed:
fix: no more linker error; seperated printCANframe()

ds196 committed:
refactor: replace Can0 with ESP32Can

ds196 committed:
feat: (by Charles) add pins for Arm

ds196 committed:
feat: update vic can implementation to add extra bit to cmd id

ds196 committed:
refactor: consolidate response code

ds196 committed:
feat: add vic can relay, cleanup code

ds196 committed:
feat: add method to toggle viccan relay

ds196 committed:
feat: add LSS macros to DIGIT

ds196 committed:
feat: finish adding viccan misc phys ctrl commands

Charles-Marmann committed:
Changed the motor amount constant for both main and motor mcu

ds196 committed:
feat: add debug output to VicCan

ds196 committed:
feat: add more debug output

ds196 committed:
feat: MOAR DEBUG

ds196 committed:
fix: remove timeout to make CAN work

ds196 committed:
feat: add submodule-specific commands to VicCan

ds196 committed:
feat: remove CMD_TEST VicCan command

Prevents a CAN ID with all zeroes

ds196 committed:
fix: frames with no data are now encoded correctly

When there is no data, instead of canData[0] == 0, now canData.size() == 0 because dlc = 0

ds196 committed:
feat: VicCan can_relay_fromvic now parses data before relaying it

No need for the raw CAN data...
Also, numerical precision is selectable with the FEEDBACK_PRECISION macro.

ds196 committed:
style: rename something() to sendFrame()

ds196 committed:
feat: relayOn/Off() now notifies Serial

ds196 committed:
feat: use MCU names rather than IDs for Serial

Also update standard to change broadcast ID from 0 to 1 to use toInt() to detect ID vs name used in relayFromSerial()

ds196 committed:
feat: add MCU name to relayOn() message

Charles-Marmann committed:
Added Arm Manual Control and Motor Feedback IDs

Charles-Marmann committed:
Axis Zero Pin Added

Charles-Marmann committed:
add Turn right and left constants

ds196 committed:
fix: remove convertControllerValue() from setDuty()

Charles-Marmann committed:
Properly Assign The Pins For Arm

ds196 committed:
fix: rename map() to map_d() to fix name conflict; revert setDuty()

ds196 committed:
feat: sync VicCAN commands with spreadsheet

Add arm manual control and REV motor feedback

ds196 committed:
fix: linker is a whiney bitch

ds196 committed:
refactor: remove OLD_ASTRACAN_ENABLE macro

ds196 committed:
fix: redundant VicCAN command

CMD_DIGIT_EF_CTRL is handled by CMD_DCMOTOR_CTRL; move and rename to CMD_DIGIT_WRIST_ROLL for manual wrist control

ds196 committed:
refactor: remove Adafruit Neopixel dependency from AstraNP and add warning error

ds196 committed:
style: comment the shit out of AstraVicCAN.h

ds196 committed:
refactor: remove deprecated AstraCAN for ESP32

Replaced with already-defined ESP32Can

ds196 committed:
refactor: resolve TODO comments

ds196 committed:
fix: finish deprecating `TwaiCAN AstraCAN`

ds196 committed:
feat: add CITADEL motor mcu pins

ds196 committed:
feat!: add citadel fan command, change arm command ID

ds196 committed:
feat: add CITADEL pins

ds196 committed:
refactor: move parseData() from VicCanController to VicCanFrame

ds196 committed:
refactor: move ESP32Can out of VicCanController; isolate in VicCanFrame

ds196 committed:
refactor: disable CAN functionality if using MCU without CAN

ryleu commented on examples/Template/Template.cpp:
```diff
@@ -38,7 +143,9 @@ void loop() {
+        //--------//
+        //  Misc  //
+        //--------//
         /**/ if (command == "ping") {
```
/**/ does not need to be here

ds196 replied:
Was just added for readability, but I supoooose it is unnecessary

ryleu commented on include/AstraMisc.h:
```diff
@@ -27,12 +50,345 @@
 // void loopHeartbeatsNew(AstraMotors* motors[], const int numMotors);
```
should be implemented or removed

ryleu commented on include/AstraMisc.h:
```diff
@@ -27,12 +50,345 @@
+
+
+#ifdef ARDUINO_RASPBERRY_PI_PICO
+unsigned hwPinToGPIO(unsigned hwPin) {
```
devious uncommented block

ryleu commented on include/AstraMotors.h:
```diff
@@ -8,44 +8,153 @@
-    AstraMotors(int setMotorID, int setCtrlMode, bool inv, int setMaxSpeed, float setMaxDuty);
-    float convertControllerValue(
-        float stickvalue);  // Converts the joystick value -1 < 0 < 1 to 1700 < 1500 < 1300
-    // void setMotorMultiplier(float val);                 // Set the speedMultiplier variable
```
should be implemented or removed

ds196 replied:
Removed afaik

ryleu commented on include/AstraNP.h:
```diff
@@ -0,0 +1,50 @@
+
+#include <Arduino.h>
+
+#error "Do not use AstraNP.h yet!!! This library is 100% WIP and is not ready for use at all..."
```
☠️

ds196 replied:
Lmao I suppose we can keep it out of main for now

ryleu commented on include/AstraREVCAN.h:
```diff
@@ -0,0 +1,321 @@
+    kBool = 0x3,
+};
+
+enum class sparkMax_ConfigParameter : int32_t {
```
this is devious. might be worth putting in a separate file?

ds196 replied:
Yeah, good point... that new file is 244 lines :sob:

ryleu commented on include/AstraVicCAN.h:
```diff
@@ -0,0 +1,682 @@
+// How many decimal places to include in data from rover
+#define FEEDBACK_PRECISION 7
+
+#if (defined(ESP32) && __has_include("ESP32-TWAI-CAN.hpp"))  // || (defined(CORE_TEENSY) && __has_include("FlexCAN_T4.h"))
```
commented out goobers!!

ryleu commented on include/AstraVicCAN.h:
```diff
@@ -0,0 +1,682 @@
+#define DT_NA CanDataType::DT_8i8
+
+// Command IDs for VicCAN frames
+/*typedef*/ enum CanCmdId : uint8_t {
```
commented out typedef

ryleu commented on include/AstraVicCAN.h:
```diff
@@ -0,0 +1,682 @@
+    if (str.length() == 0)
+        return false;
+    
+    /**/ if (str == "broadcast")
```
why do you do this

ryleu commented on include/AstraVicCAN.h:
```diff
@@ -0,0 +1,682 @@
+ * @return String containing only the name of the MCU
+ */
+String mcuIdToString(const McuId mcuID) {
+    /**/ if (mcuID == McuId::MCU_BROADCAST)
```
again bruh with the /**/

ryleu commented on include/AstraVicCAN.h:
```diff
@@ -0,0 +1,682 @@
+        clear();
+    }
+
+    // Resets all data values; for use with static keyword
```
turning all instances of regular comments on functions into documentation comments (the ones that start with `/**`) might be a good idea. if you need me to, i can point out every place where this needs to happen

ryleu commented on include/AstraVicCAN.h:
```diff
@@ -0,0 +1,682 @@
+        // 64-bit memory address. re-interpret_cast is used to accomplish this as C++ is
+        // not going to mess with an unsigned int.
+
+        /**/ if (dlc == 0) {
```
/**/

ryleu commented on include/AstraVicCAN.h:
```diff
@@ -0,0 +1,682 @@
+        /**/ if (dlc == 0) {
+            // No data to parse
+        }
+        else if (dataType == CanDataType::DT_1f64) {  // 1 double
```
should comment bitwise operations like this

ryleu commented on include/AstraVicCAN.h:
```diff
@@ -0,0 +1,682 @@
+        res.reserve(30);
+        res += mcuIdToString(mcuId);
+        res += " [";
+        res += cmdId;
```
would probably look better and be easier to read as `res += "[" + cmdId + "]";`

ds196 replied:
Just trying to avoid Arduino String fuckiness https://docs.arduino.cc/built-in-examples/strings/StringAdditionOperator/

ryleu replied:
hateful, but you gotta do what you gotta do

ryleu commented on src/AstraREVCAN.cpp:
```diff
@@ -0,0 +1,218 @@
+
+    ESP32Can.writeFrame(msg);
+
+    // WARNING: DO NOT USE THIS CODE. THE MOTORS WILL HAVE A SEIZURE.
```
whar ?

would be good to figure this out. comments like this are okay, but like bruh

ds196 replied:
*Hours* were spent on this... We never could figure out why using functions instead of the raw code would make the motors completely shit themselves

ryleu replied:
that is very odd... i guess that's fine for now and we can fix it when we get the chance. wonder if using a higher optimization level would help?

ryleu reviewed (changes requested):
that is a big merge request

ds196 committed:
style: resolve some of Riley's comments

ds196 committed:
refactor: keep AstraNP out of main

Not usable yet lmao

ds196 committed:
style: resolve the rest of Riley's comments

ds196 committed:
feat: consolidate REVCAN control commands, add current control

TristanMcGinnis commented on include/AstraCAN.h:
```diff
@@ -1,38 +1,55 @@
 
-#if !__has_include("FlexCAN_T4.h")
+// Must have the CAN library corresponding to the current MCU
+#if defined(CORE_TEENSY) && !__has_include("FlexCAN_T4.h")  // If on Teensy, must have FlexCAN_T4
```
Honestly we could just remove the teensy stuff, since we've now moved away from it entirely on every module.

TristanMcGinnis commented on include/project/TESTBED.h:
Think that we need this file? I know wheel circumference *might* be used in the future but otherwise core and testbed *should* remain identical. They do use the same PCB now.

TristanMcGinnis reviewed (approved):
Left 2 comments. Not critical things, everything else looks good to me.

ds196 committed:
feat: add sendSpeed() to bypass acceleration for velocity control

ryleu commented on include/AstraMisc.h:
```diff
@@ -19,13 +19,99 @@
+
+
+/**
+ * @brief 
```
missing description

ryleu commented on include/AstraREVTypes.h:
```diff
@@ -0,0 +1,244 @@
+/**
+ * @file AstraREVTypes.h
+ * @author David Sharpe (ds0196@uah.edu)
+ * @brief 
```
missing description

ryleu commented on include/AstraMisc.h:
```diff
@@ -19,13 +19,99 @@
+    // Protection against infinite loop
+    unsigned count = 0;
+    while (count++,
+           count < 200 /*arbitrary limit on number of delims because while(true) is scary*/) {
```
this seems off to me... if we get a command > 200 delims, that command would still get executed even though it's not fully processed 

i think the output string should be set to some sort of NOOP command to prevent this

ryleu commented on src/AstraMotors.cpp:
```diff
@@ -5,50 +5,49 @@
-
-    currentDutyCycle = 0;
-    setDutyCycle = 0;
+AstraMotors::AstraMotors(int setMotorID, sparkMax_ctrlType setCtrlMode, bool inv,
```
this is totally a nitpick, but inv should be spelled out as "invert" or "setInverted" or something along those lines. reduces cognitive load that way, which is a must for maintainable code

ryleu commented on src/AstraMotors.cpp:
```diff
@@ -5,50 +5,49 @@
+    gearBox = 64;  // default to 64:1 for Testbed
+#   else
+    gearBox = 100;  // default to 100:1 for Core
+#   endif
```
this isn't a good way to do this. gearbox ratio should be passed in from the piece of code that instantiates the class

otherwise, if you need more motors later with different gearboxes, you need to edit your library

ds196 committed:
refactor!: remove `#ifdef ARM` statements

Arm's main.cpp will need to be edited to replace `setDuty()` with `sendDuty()` to disable acceleration

ds196 committed:
feat: sync vicCAN commands with spreadsheet

ds196 committed:
feat: add parseREVParameter()

ds196 committed:
feat: add AstraMotors::parseStatus() to replace seperate methods for each apiId

ds196 committed:
refactor: add MotorOptions for AstraMotors constructor

ds196 committed:
style: move printREVFrame() from AstraMotors to AstraREVCAN

Also fix linker error with defaultMotorOpts

ds196 committed:
refactor: abort MotorOptions

Doesn't work how I thought it would...

ds196 committed:
fix: vicCAN will always relay to serial if no can available

ryleu commented:
lgtm

ds196 committed:
fix: VicCAN self-relay now parses data correctly

ryleu reviewed (approved):
i am a HUGE FAN of this change. i love how instead of it working like ass and being broken, it works good and isn't broken!

ds196 committed:
feat: add new Digit board pins

ds196 committed:
feat: enable overriding viccan float precision

ds196 committed:
refactor: move covertADC() into AstraMisc.h

ds196 committed:
feat: add BNO and BMP commands to VicCAN

ds196 committed:
style: final commit before merge

ds196 commented:
**204 commits**

ds196 commented:
Arm, Core, and Digit have been running fairly consistently aside from my occasional fuck-up in the corresponding main.cpp, so I am going to merge. Bugs will be fixed as they come up.
lgtm
